### ISSUE TITLE
Lint warning when calling static methods on subclasses

### ISSUE THREAD

**Issue by shicks** [google/closure-compiler#1889]
We should consider having `Es6ToEs3Converter` add `@deprecated` to any static methods copied from base class to subclass, so that the following produces a lint warning:

```js
class Base { static foo() {} }
class Sub extends Base {}
Sub.foo();
```

**Comment by brad4d** [comment]
Why shouldn't one call an inherited static method from a child class?

**Comment by MatrixFrog** [comment]
This came from a style-guide discussion yesterday. There are basically two reasons:

1. It's bad style (at least in my opinion). If it's a static method it should be called statically. If you call x.foo() and you know that x is some class but you don't know (at compile time) exactly which one, that's not really a "static" call.
2. It can cause problems with compiler renaming, [as described here](https://developers.google.com/closure/compiler/docs/limitations#implications-of-object-property-flattening) ("Using static methods without knowing which class they're called on")

#1 is just a matter of opinion, and #2 may be fixed eventually by changing how CollapseProperties works in some way. So this is not necessarily set in stone but it seems to be reasonable right now. We'll have to keep an eye on the ongoing style discussions too.

**Comment by brad4d** [comment]
The static method is being called statically in this example. It's being
called directly on a class name, "Sub", not on a variable containing a
class.

class Base { static foo() {} }class Sub extends Base {}Sub.foo();

I don't think this code should cause any kind of warning. If it does,
we're basically saying you must always pretend like subclasses do not
inherit static methods.

**Comment by MatrixFrog** [comment]
I am saying that, thank you for stating it more clearly than I did :) I think static inheritance is a bad feature.

**Comment by brad4d** [comment]
Why is it a bad feature? It's exactly what I would expect to happen,
because it's consistent with other OO languages like Java and C++. I would
find *not* inheriting static methods very surprising behavior.

**Comment by shicks** [comment]
Yes, but Java also emits a warning when you call `Sub.foo()`, since there's no reason to call that over `Base.foo()`.

**Comment by ChadKillingsworth** [comment]
It does prevent you from referencing the super class name in more than one place - which I believe was a goal of the `class` keyword syntax.
